packages/plugin-constraints/sources/commands/constraints/query.ts

```diff
@@ -0,0 +1,45 @@
+import {Configuration, Project, PluginConfiguration} from '@berry/core';
+import {Writable}                                    from 'stream';
+
+import {Constraints}                                 from '../../Constraints';
+
+function valueToString(value: string|null): string {
+  if (typeof value !== `string`)
+    return `${value}`;
+
+  if (value.match(/^[a-zA-Z][a-zA-Z0-9_]+$/))
+    return value;
+
+  return `'${value}'`;
+}
+
+// eslint-disable-next-line arca/no-default-export
+export default (clipanion: any, pluginConfiguration: PluginConfiguration) => clipanion
+
+  .command(`constraints query <query>`)
+
+  .categorize(`Constraints-related commands`)
+  .describe(`query the constraints`)
+
+  .detail(`
+    This command will output all matches to the given prolog query
+  `)
+
+  .example(
+    `Find all fixable invalid dependencies`,
+    `yarn constraints query 'workspace(WorkspaceCwd), dependency_type(DependencyType), gen_enforced_dependency_range(WorkspaceCwd, DependencyName, DependencyRange, DependencyType).'`,
+  )
+
+  .action(async ({cwd, stdout, query}: {cwd: string, stdout: Writable, query: string}) => {
+    const configuration = await Configuration.find(cwd, pluginConfiguration);
+    const {project} = await Project.find(configuration, cwd);
+    const constraints = await Constraints.find(project);
+    
+    for (const result of await constraints.query(query)) {
+      const stringifiedResult = Array.from(Object.entries(result), ([variable, value]) => `${variable} = ${valueToString(value)}`).join(`,\n`);
```

Comment:
I'd change the output to:

- Use `StreamReport` (by consistency with the other commands)
- Put all variables for a result on the same line (the logs in general have one line by message, useful to count)
- Add colors on the variable and values to make it easier for the brain to parse

So in the end it would become:

```
BR0001 WorkspaceCwd=..., DependencyName=lodash, DependencyType=Prod
BR0001 WorkspaceCwd=..., DependencyName=webpack, DependencyType=Prod
```

And then later I think I'll make a diff to add JSON output (I want to rework a bit the JSON reporter first - it currently is a bit weirder to use than I'd like).

Reply:
I went with output in the same style as swipl, but I can change it if you prefer.

Putting everything in one line won't help with readability imo, especially given that the `WorkspaceCwd` can get quite long:

```bash
$ y constraints query 'workspace(WorkspaceCwd), dependency_type(DependencyType), gen_enforced_dependency_range(WorkspaceCwd, DependencyName, DependencyRange, DependencyType).'
# ...
WorkspaceCwd = '/Volumes/Workspaces/private/berry/packages/acceptance-tests/pkg-tests-specs', DependencyType = dependencies, DependencyName = 'pkg-tests-core', DependencyRange = 'workspace:*'
```

Once the terminal starts wrapping the lines, it quickly becomes a complex to read

Reply:
Fair point - what do you think of this then? Unicode box characters to group the entries together:

```
BR0001 ┌ WorkspaceCwd    = '/Volumes/Workspaces/private/berry/packages/acceptance-tests/pkg-tests-specs'
BR0001 │ DependencyIdent = 'pkg-tests-core'
BR0001 │ DependencyRange = 'workspace:*'
BR0001 └ DependencyType  = dependencies
BR0001 ┌ WorkspaceCwd    = '/Volumes/Workspaces/private/berry/packages/acceptance-tests/pkg-tests-specs'
BR0001 │ DependencyIdent = 'pkg-tests-core'
BR0001 │ DependencyRange = 'workspace:*'
BR0001 └ DependencyType  = dependencies
```

As a followup later the BR0001 could be replaced by the entry number (like `#00001` then `#00002` etc), but for a first iteration that shouldn't be a problem.

Reply:
I'm not quite sure how to implement this with the current reports. Do I just build the unicode box in the command itself?

Reply:
Yup! I think they should be supported pretty much everywhere nowadays.